mllib/src/main/scala/org/apache/spark/mllib/linalg/distributed/BlockMatrix.scala

```diff
@@ -0,0 +1,242 @@
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.spark.mllib.linalg.distributed
+
+import breeze.linalg.{DenseMatrix => BDM}
+
+import org.apache.spark.{Logging, Partitioner}
+import org.apache.spark.mllib.linalg._
+import org.apache.spark.mllib.rdd.RDDFunctions._
+import org.apache.spark.rdd.RDD
+import org.apache.spark.storage.StorageLevel
+
+/**
+ * A grid partitioner, which stores every block in a separate partition.
+ *
+ * @param numRowBlocks Number of blocks that form the rows of the matrix.
+ * @param numColBlocks Number of blocks that form the columns of the matrix.
+ */
+private[mllib] class GridPartitioner(
+    val numRowBlocks: Int,
+    val numColBlocks: Int,
+    val numParts: Int) extends Partitioner {
+  // Having the number of partitions greater than the number of sub matrices does not help
+  override val numPartitions = math.min(numParts, numRowBlocks * numColBlocks)
+
+  /**
+   * Returns the index of the partition the SubMatrix belongs to. Tries to achieve block wise
+   * partitioning.
+   *
+   * @param key The key for the SubMatrix. Can be its position in the grid (its column major index)
+   *            or a tuple of three integers that are the final row index after the multiplication,
+   *            the index of the block to multiply with, and the final column index after the
+   *            multiplication.
+   * @return The index of the partition, which the SubMatrix belongs to.
+   */
+  override def getPartition(key: Any): Int = {
+    key match {
+      case (blockRowIndex: Int, blockColIndex: Int) =>
+        getBlockId(blockRowIndex, blockColIndex)
+      case (blockRowIndex: Int, innerIndex: Int, blockColIndex: Int) =>
+        getBlockId(blockRowIndex, blockColIndex)
+      case _ =>
+        throw new IllegalArgumentException(s"Unrecognized key. key: $key")
+    }
+  }
+
+  /** Partitions sub-matrices as blocks with neighboring sub-matrices. */
+  private def getBlockId(blockRowIndex: Int, blockColIndex: Int): Int = {
+    val totalBlocks = numRowBlocks * numColBlocks
+    // Gives the number of blocks that need to be in each partition
+    val partitionRatio = math.ceil(totalBlocks * 1.0 / numPartitions).toInt
+    // Number of neighboring blocks to take in each row
+    val subBlocksPerRow = math.ceil(numRowBlocks * 1.0 / partitionRatio).toInt
```

Comment:
This is wrong. If we have 10x20 blocks and 10 partitions. `partitionRatio = 20` and `subBlocksPerRow = 1` and `subBlocksPerCol = 1`. Then we will have 200 partitions. Please update the assignment login and add tests. Btw, it may be worth computing `subBlocksPerRow` and `subBlocksPerCol` in constructor. I would recommend renaming them to `numRowBlocksPerPartition` and `numColBlocksPerPartition`.